File: algorithms/bfs.py
from algorithms import algorithm
from collections import deque

class BFS(algorithm.Algorithm):
    def __init__(self, graph, start_node, end_node):
        super().__init__(graph, start_node, end_node)
        self.queue = deque([self.start_node])
        self.visited = {self.start_node}
        self.came_from = {self.start_node: None}

    def solve(self):
        while self.queue:
            current_node = self.queue.popleft()

            if current_node == self.end_node:
                return self.reconstruct_path()

            for neighbor in range(len(self.graph[current_node])):
                if self.graph[current_node][neighbor] > 0 and neighbor not in self.visited:
                    self.visited.add(neighbor)
                    self.came_from[neighbor] = current_node
                    self.queue.append(neighbor)
        
        return None

    def reconstruct_path(self):
        path = []
        current_node = self.end_node
        while current_node is not None:
            path.append(current_node)
            current_node = self.came_from[current_node]
        path.reverse()
        return path

File: src/app/core/algorithms/bfs.ts
import { Algorithm, NodeState } from './algorithm';

export class BFS extends Algorithm {
  private queue: number[] = [];
  private visitedSet: Set<number> = new Set();
  private frontierSet: Set<number> = new Set();
  private cameFrom: Map<number, number | null> = new Map();

  constructor(graph: number[][], startNode: number, endNode: number) {
    super(graph, startNode, endNode);
    this.queue.push(this.startNode);
    this.visitedSet.add(this.startNode);
    this.cameFrom.set(this.startNode, null);
  }

  *solve(): Generator<NodeState, void, unknown> {
    this.frontierSet.add(this.startNode);
    yield {
      frontier: new Set(this.frontierSet),
      visited: new Set(this.visitedSet),
      currentNode: this.startNode,
      path: [this.startNode]
    };

    while (this.queue.length > 0) {
      const currentNode = this.queue.shift()!;
      this.frontierSet.delete(currentNode);
      this.visitedSet.add(currentNode);

      if (currentNode === this.endNode) {
        yield {
          frontier: new Set(this.frontierSet),
          visited: new Set(this.visitedSet),
          currentNode: currentNode,
          path: this.reconstructPath(currentNode)
        };
        return;
      }

      for (let neighbor = 0; neighbor < this.graph[currentNode].length; neighbor++) {
        const weight = this.graph[currentNode][neighbor];
        if (weight > 0 && !this.visitedSet.has(neighbor) && !this.frontierSet.has(neighbor)) {
          this.frontierSet.add(neighbor);
          this.cameFrom.set(neighbor, currentNode);
          this.queue.push(neighbor);
        }
      }

      yield {
        frontier: new Set(this.frontierSet),
        visited: new Set(this.visitedSet),
        currentNode: currentNode,
        path: this.reconstructPath(currentNode)
      };
    }
  }

  private reconstructPath(targetNode: number): number[] {
    const path: number[] = [];
    let current: number | null = targetNode;
    while (current !== null) {
      path.push(current);
      current = this.cameFrom.get(current) ?? null;
    }
    return path.reverse();
  }
}
